connectors/src/connectors/webcrawler/temporal/activities.ts

```diff
@@ -0,0 +1,156 @@
+import { ModelId } from "@dust-tt/types";
+import { CheerioCrawler, Configuration } from "crawlee";
+import turndown from "turndown";
+
+import { dataSourceConfigFromConnector } from "@connectors/lib/api/data_source_config";
+import { upsertToDatasource } from "@connectors/lib/data_sources";
+import { Connector } from "@connectors/lib/models";
+import {
+  WebCrawlerConfiguration,
+  WebCrawlerFolder,
+} from "@connectors/lib/models/webcrawler";
+import {
+  reportInitialSyncProgress,
+  syncSucceeded,
+} from "@connectors/lib/sync_status";
+
+export async function crawlWebsiteByConnectorId(connectorId: ModelId) {
+  const connector = await Connector.findByPk(connectorId);
+  if (!connector) {
+    throw new Error(`Connector ${connectorId} not found.`);
+  }
+  const webCrawlerConfig = await WebCrawlerConfiguration.findOne({
+    where: {
+      connectorId,
+    },
+  });
+  if (!webCrawlerConfig) {
+    throw new Error(`Webcrawler configuration not found for connector.`);
+  }
+
+  const dataSourceConfig = dataSourceConfigFromConnector(connector);
+  let pageCount = 0;
+  let errorCount = 0;
+  const createdFolders = new Set<string>();
+
+  const crawler = new CheerioCrawler(
+    {
+      maxRequestsPerCrawl: 300,
+      maxConcurrency: 10,
+
+      async requestHandler({ $, request, enqueueLinks }) {
+        const extracted = new turndown()
+          .remove(["style", "script", "iframe"])
+          .turndown($.html());
+
+        const pageTitle = $("title").text();
+
+        await enqueueLinks();
+
+        const folders = getAllFoldersForUrl(request.url);
+        for (const folder of folders) {
+          if (!folder.startsWith(webCrawlerConfig.url)) {
+            continue;
+          }
+          if (createdFolders.has(folder)) {
+            continue;
+          }
+
+          const logicalParent =
+            folder === webCrawlerConfig.url ? null : getFolderForUrl(folder);
+          await WebCrawlerFolder.upsert({
+            url: folder,
+            parentUrl: logicalParent,
+            connectorId: connector.id,
+            webcrawlerConfigurationId: webCrawlerConfig.id,
+            ressourceType: "folder",
+            title: null,
+          });
+          createdFolders.add(folder);
+        }
+
+        const logicalParent =
+          request.url === webCrawlerConfig.url
+            ? null
+            : getFolderForUrl(request.url);
+        const updatedFolder = await WebCrawlerFolder.upsert({
+          url: request.url,
+          parentUrl: logicalParent,
+          connectorId: connector.id,
+          webcrawlerConfigurationId: webCrawlerConfig.id,
+          ressourceType: "file",
+          title: pageTitle,
+        });
+
+        await upsertToDatasource({
+          dataSourceConfig,
+          documentId: updatedFolder[0].id.toString(),
```

Comment:
this is the ModelId ? This assumes that you will forever upsert correctly which is a foot gun IMHO. We likely want hashed IDs to ease incremental syncing?

Reply:
Also there is conflation of folders and pages which will make it hard for people to reason about this. 

You won't be able to distinguish a folder from a page in the getPermission UI will you?

Reply:
There is a unique key on `WebCrawlerFolder.{configurationId, url, ressourceType}`, so we can differentiate a page and a folder with the same URL using `ressourceType` and the primary would be different. 

Agree that this is not ideal, but this guarantees no collision, no need for locking (DB does it for us).
Happy to use a sha256 hash instead.

Reply:
I agree it works. But I believe it's going to be a hell to maintain and extend.